src/sundials/sundials_adaptcontroller.c

```diff
@@ -0,0 +1,261 @@
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2023, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------
+ * This is the implementation file for a generic SUNAdaptController
+ * package. It contains the implementation of the SUNAdaptController
+ * operations listed in sundials_adaptcontroller.h
+ * -----------------------------------------------------------------*/
+
+#include <sundials/sundials_adaptcontroller.h>
+
+/* -----------------------------------------------------------------
+ * Create a new empty SUNAdaptController object
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController SUNAdaptController_NewEmpty(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptController_Ops ops;
+
+  /* a context is required */
+  if (sunctx == NULL) return(NULL);
+
+  /* create controller object */
+  C = NULL;
+  C = (SUNAdaptController) malloc(sizeof *C);
+  if (C == NULL) return(NULL);
+
+  /* create matrix ops structure */
+  ops = NULL;
+  ops = (SUNAdaptController_Ops) malloc(sizeof *ops);
+  if (ops == NULL) { free(C); return(NULL); }
+
+  /* initialize operations to NULL */
+  ops->gettype               = NULL;
+  ops->destroy               = NULL;
+  ops->reset                 = NULL;
+  ops->estimatestep          = NULL;
+  ops->estimatestepandorder  = NULL;
+  ops->estimatemristeps      = NULL;
+  ops->estimatesteptol       = NULL;
+  ops->setdefaults           = NULL;
+  ops->write                 = NULL;
+  ops->setmethodorder        = NULL;
+  ops->adjustcontrollerorder = NULL;
+  ops->seterrorbias          = NULL;
+  ops->update                = NULL;
+  ops->updatemrih            = NULL;
+  ops->updatemritol          = NULL;
+  ops->space                 = NULL;
+
+  /* attach ops and initialize content to NULL */
+  C->ops     = ops;
+  C->content = NULL;
+  C->sunctx  = sunctx;
+
+  return(C);
+}
+
+
+/* -----------------------------------------------------------------
+ * Free a generic SUNAdaptController (assumes content is already empty)
+ * ----------------------------------------------------------------- */
+
+void SUNAdaptController_FreeEmpty(SUNAdaptController C)
+{
+  if (C == NULL)  return;
+
+  /* free non-NULL ops structure */
+  if (C->ops)  free(C->ops);
+  C->ops = NULL;
+
+  /* free overall SUNAdaptController object and return */
+  free(C);
+  return;
+}
+
+
+/* -----------------------------------------------------------------
+ * Required functions in the 'ops' structure for non-NULL controller
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType(SUNAdaptController C)
+{
+  if (C == NULL) { return SUN_ADAPTCONTROLLER_NONE; }
```

Comment:
I think we need to be very careful about purposefully making `NULL` to be a valid object, which is what we seem to be doing here.  This could lead to some hard to debug issues I imagine. I think the better approach would be to implement a controller that does not change the step size (as I recommended in a different comment).